### PR TITLE
8255: Export Products action doesn't consider hide_for_product_page value.

### PR DESCRIPTION
**Pull request by nmalevanec**
### Description
Fix Export and Import Products action doesn't consider hide_for_product_page value.

### Fixed Issues (if relevant)

1. magento/magento2#8255: Export Products action doesn't consider hide_for_product_page value

### Manual testing scenarios
1. Create a simple product and upload an image.
2. Change the scope to 1 store view.
3. Set the image as Hide for product page in that store view.
4. Navigate to System -> Export.
5. Select Products and then Continue.
6. Check that CSV files should have 2 lines. The default values at the first row of values.
7. The second lines should contain values on sku, store_view_code and hide_from_product_page columns.
8. Remove created simple product.
9. Navigate to System -> Import.
10. Import product from exported CSV with add/update behavior.
11. Check, imported product at 1 store view has "Hide fro product page" checked for image.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by dmanners** [comment]
@nmalevanec could you have a look at the merge conflicts for me please.

**Comment by nmalevanec** [comment]
@dmanners Done.

File: app/code/Magento/CatalogImportExport/Model/Export/Product.php [hunk 1]
```diff
@@ -1024,12 +1026,10 @@ protected function collectRawData()
                     unset($data[$itemId][$storeId][self::COL_ADDITIONAL_ATTRIBUTES]);
                 }
 
-                if (!empty($data[$itemId][$storeId]) || $this->hasMultiselectData($item, $storeId)) {
-                    $attrSetId = $item->getAttributeSetId();
-                    $data[$itemId][$storeId][self::COL_STORE] = $storeCode;
-                    $data[$itemId][$storeId][self::COL_ATTR_SET] = $this->_attrSetIdToName[$attrSetId];
-                    $data[$itemId][$storeId][self::COL_TYPE] = $item->getTypeId();
-                }
+                $attrSetId = $item->getAttributeSetId();
```
Comment (dmanners) [inline app/code/Magento/CatalogImportExport/Model/Export/Product.php:1029]: Does this also change the behavior for non-image multi-select data also? Do we have a test covering this?

File: app/code/Magento/CatalogImportExport/Model/Import/Product.php [hunk 2]
```diff
@@ -2972,4 +2991,64 @@ private function getExistingSku($sku)
...
         return $this->_oldSku[strtolower($sku)];
     }
+
+    /**
+     * Set product images 'disable' = 0 for specified store.
+     *
+     * @param array $mediaGalleryData
+     * @return array
+     */
+    private function restoreDisableImage(array $mediaGalleryData)
```
Comment (dmanners) [inline app/code/Magento/CatalogImportExport/Model/Import/Product.php]: I know we already have a massive file here but would it make sense/be possible to extract the image processing to a separate class?

File: app/code/Magento/CatalogImportExport/Model/Import/Product.php [hunk 3]
```diff
@@ -2972,4 +2991,64 @@ private function getExistingSku($sku)
...
+     * Remove existed images from insert data.
+     *
+     * @param array $multiInsertData
+     * @param array $imageNames
+     * @return array
+     */
+    private function filterImageInsertData(array $multiInsertData, array $imageNames)
+    {
+        //Remove image duplicates for stores.
+        $multiInsertData = array_map("unserialize", array_unique(array_map("serialize", $multiInsertData)));
```
Comment (dmanners) [inline app/code/Magento/CatalogImportExport/Model/Import/Product.php]: Is this format json? I would be a bit worried that we are calling unserialize/serialize on data that we do not trust/have control over.

**Review by dmanners — CHANGES_REQUESTED** [review]

**Comment by dmanners** [comment]
There are two test failures currently. Could you look into them please.

```

1) Magento\CatalogImportExport\Model\Export\ProductTest::testExportWithMedia
--
13-Nov-2017 04:15:27 | [exec] PDOException: SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`bamboo_integration_0`.`pfx_catalog_product_option_type_price`, CONSTRAINT `FK_0404E60DE3B8ACE3BC2CE062A74D0A97` FOREIGN KEY (`option_type_id`) REFERENCES `pfx_catalog_product_option_type_valu)


2) Magento\CatalogImportExport\Model\Export\ProductTest::testExportWithMedia
--
13-Nov-2017 04:15:27 | [exec] Failed asserting that Array &0 (
13-Nov-2017 04:15:27 | [exec]     'additional_images' => 'additional_images'
13-Nov-2017 04:15:27 | [exec]     '/m/a/magento_image.jpg' => '/m/a/magento_image.jpg'
13-Nov-2017 04:15:27 | [exec]     '' => ''
13-Nov-2017 04:15:27 | [exec] ) is identical to Array &0 (
13-Nov-2017 04:15:27 | [exec]     'hide_from_product_page' => 'hide_from_product_page'
13-Nov-2017 04:15:27 | [exec]     '' => ''
13-Nov-2017 04:15:27 | [exec]     '/m/a/magento_image.jpg' => '/m/a/magento_image.jpg'
13-Nov-2017 04:15:27 | [exec] ).
```

File: app/code/Magento/CatalogImportExport/Model/Export/Product.php (hunk 1 above)
Comment (nmalevanec) [reply app/code/Magento/CatalogImportExport/Model/Export/Product.php:1029]: It covers by Magento/CatalogImportExport/Model/Export/ProductTest::testExport() for instance. Exported product has multiselect data.

File: app/code/Magento/CatalogImportExport/Model/Import/Product.php (hunk 2 above)
Comment (nmalevanec) [reply app/code/Magento/CatalogImportExport/Model/Import/Product.php]: Done.

File: app/code/Magento/CatalogImportExport/Model/Import/Product.php (hunk 3 above)
Comment (nmalevanec) [reply app/code/Magento/CatalogImportExport/Model/Import/Product.php]: Removed.

**Comment by nmalevanec** [comment]
@dmanners integration test fixed.

**Comment by dmanners** [comment]
@nmalevanec just to keep you updated there is a test that compares the installation database with sample data compared to 2.2-develop and sample data. This is showing differences after your change. I am looking into this and will get back to you about it.

**Comment by nmalevanec** [comment]
@dmanners Semantic Version Checker fixed.

**Review by dmanners — APPROVED** [review]

**Comment by devgas** [comment]
@okobchenko  Fix for #12356 doesn't exist in Magento 2.2.3

**Comment by dmanners** [comment]
Hi @devgas 2.2.3 was a security patch release and as such was limited in what the contents of this release were. This should be included in the 2.2.4 release though.

**Pull request merged**